core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -432,6 +465,93 @@ private[spark] class Executor(
   }
 
   /**
+   * Supervises the killing / cancellation of a task by sending the interrupted flag, optionally
+   * sending a Thread.interrupt(), and monitoring the task until it finishes.
+   */
+  private class TaskReaper(
+      taskRunner: TaskRunner,
+      val interruptThread: Boolean)
+    extends Runnable {
+
+    private[this] val taskId: Long = taskRunner.taskId
+
+    private[this] val killPollingIntervalMs: Long =
+      conf.getTimeAsMs("spark.task.reaper.pollingInterval", "10s")
+
+    private[this] val killTimeoutMs: Long = conf.getTimeAsMs("spark.task.reaper.killTimeout", "2m")
+
+    private[this] val takeThreadDump: Boolean =
+      conf.getBoolean("spark.task.reaper.threadDump", true)
+
+    override def run(): Unit = {
+      val startTimeMs = System.currentTimeMillis()
+      def elapsedTimeMs = System.currentTimeMillis() - startTimeMs
+      def timeoutExceeded(): Boolean = killTimeoutMs > 0 && elapsedTimeMs > killTimeoutMs
+      try {
+        // Only attempt to kill the task once. If interruptThread = false then a second kill
+        // attempt would be a no-op and if interruptThread = true then it may not be safe or
+        // effective to interrupt multiple times:
+        taskRunner.kill(interruptThread = interruptThread)
+        // Monitor the killed task until it exits:
+        var finished: Boolean = false
```

Comment:
I think that my original concern was about holding `taskRunner.synchronized` while taking the thread dump. We need to synchronize from the time that we check `taskRunner.isFinished` to when we call `taskRunner.wait()` in order to avoid a race where the task finishes right before we'd begin waiting and if we don't want to synchronize the entire loop then we need to redundantly check again in the `while` condition. My goal here was to avoid doing the `isFinished` call twice.

I suppose that lock contention shouldn't really matter here, though, since this only comes into play once a task has already been marked for killing / cancellation. Let me just lift the synchronized out of the body of the loop and I think I'll be able to simplify this slightly.